### PR TITLE
TY-2075 refactor getAssets

### PR DESCRIPTION
**Pull request by Robert-Steiner**
Ticket:

- [TY-2075], [TY-2077], [TY-2091]

**Summary:**

- The `wasmModule` and` wasmScript` `AssetTypes` have been added and the` wasm<Feature>Module`, `wasm<Feature>Script` and `wasm<Feature>Snippet` have been removed.
- We can remove `wasm<Feature>Snippet` because `genesis.js` is taking care of loading it.
- Replaced the `wasmParallel` variant with the wasm features. This way team blue doesn't need to know what wasm features are necessary to choose the right wasm module. They just pass us the set of features that are available.
- ~Moved the `Feature` enum to the web package. Added a `Feature` class with a private constructor for the mobile package.~
	- ~The only issue with that solutions is that we need to disable some lints since `flutter analyze` does not respect conditional imports but uses the default import.~ (see https://github.com/xaynetwork/xayn_ai/pull/222#issuecomment-926756405)
- Removed the hardcoded snippets path. If a `snippets` folder exists, we just upload the whole folder.
- Split the `assets` array of `assets_metadata.json` into three arrays
	- `assets`  which contains the metadata of the data assets like vocab, model etc.
	- `wasm_assets`  which contains the metadata of the wasm assets like module and script
	- and `upload` which contains the `path` and `url_suffix` for each assets that need so be uploaded
	- this also simplified the code of uploading the assets in the release ci
- Removed the pinned version of `wasm-bindgen-rayon`.
- Updated `CHANGELOG.md`.

Ideas: 

- ~What about moving the `wasmSequential`  and `wasmParallel` hash map from the common package to the web package?~ (implemented)

---

`assets.dart`

```dart
/// Warning, this file is autogenerated. Don't modify this manually.
part of 'data_provider.dart';

/// Base assets that are required for both mobile and web.
///
/// The checksum is the sha256 hash of the asset.
/// To calculate the checksum run 'shasum -a 256 vocab.txt'.
final baseAssets = <AssetType, Asset>{
  AssetType.smbertVocab: Asset(
      'smbert_v0000/vocab.txt',
      Checksum(
          'ffb2398810fa977ee39e8a0103b07f9c2614f1f9825df53a2c99e71569e3451c'),
      []),
  AssetType.smbertModel: Asset(
      'smbert_v0000/smbert.onnx',
      Checksum(
          'c07653a3cc568876071de157f0fb8ccc1a302197ddad9ead069a86140e19d913'),
      [
        Fragment(
            'smbert_v0000/smbert.onnx_11MB_00',
            Checksum(
                '7f4d6242ef88ab6daaf1e1fa1f030345c4f4046585cd9576112aa5391f42a000')),
        Fragment(
            'smbert_v0000/smbert.onnx_11MB_01',
            Checksum(
                'bd642e69234c28b3b2c25138f997505195d43a734668ce55de7adebea691ea99')),
        Fragment(
            'smbert_v0000/smbert.onnx_11MB_02',
            Checksum(
                '6987dc3baddff7b742d7a481209b2a4d30097abf911fa29d50bc271dec121008')),
        Fragment(
            'smbert_v0000/smbert.onnx_11MB_03',
            Checksum(
                '510ff0115aef91497fc5ef8ffcfd4c7646644d63ddc0e0635d765cf8f19b8b36')),
        Fragment(
            'smbert_v0000/smbert.onnx_11MB_04',
            Checksum(
                '52e75ac8310bead0817a4948c126d2c2105a2213ec72d4f6864fad15d07d3491')),
      ]),
  AssetType.qambertVocab: Asset(
      'qambert_v0001/vocab.txt',
      Checksum(
          'ffb2398810fa977ee39e8a0103b07f9c2614f1f9825df53a2c99e71569e3451c'),
      []),
  AssetType.qambertModel: Asset(
      'qambert_v0001/qambert.onnx',
      Checksum(
          '030e7d68cc82f59640c6b989f145d6eaa7609c8615bd6ae9e7890e1a68be5b71'),
      [
        Fragment(
            'qambert_v0001/qambert.onnx_11MB_00',
            Checksum(
                '864c75b3132c6282ff2e266ec0e51972371fae60b1678fcb90408e92eb06dfb5')),
        Fragment(
            'qambert_v0001/qambert.onnx_11MB_01',
            Checksum(
                '36d3e3162e60fefd563e7a857a74d03755ed899dc79369f3083fa5ca33c9fed5')),
        Fragment(
            'qambert_v0001/qambert.onnx_11MB_02',
            Checksum(
                '7db90e4fe94690593115032d2b1a30b7b76a2a6b5481cabcc7ffab6b972a76ce')),
        Fragment(
            'qambert_v0001/qambert.onnx_11MB_03',
            Checksum(
                'b0c84da6e10a8ae5c0a1470e5144a4ab5dd142342c25d12968e76f72fe17c32f')),
        Fragment(
            'qambert_v0001/qambert.onnx_11MB_04',
            Checksum(
                '510e9cf989dfc50fe64348d715a543006e04f9e39d75e9ba636e139137f29ac5')),
      ]),
  AssetType.ltrModel: Asset(
      'ltr_v0000/ltr.binparams',
      Checksum(
          'f36a874ba26ea154a02070180cfcabaf191f654ecff5bb94a8472d2881d4c062'),
      []),
};

enum AssetType {
  smbertVocab,
  smbertModel,
  qambertVocab,
  qambertModel,
  ltrModel,
  wasmModule,
  wasmScript,
}

final wasmSequential = <AssetType, Asset>{
  AssetType.wasmModule: Asset(
      'wasm_sequential_bd59057417018bb747b937c7a9515dc7b0af8f6807c0d6d4a2097ba0d9ec0f5d/genesis_bg.wasm',
      Checksum(
          '3699f24f4e65a5129ee00a6274f8011e2f4d6a4b8546141399be8daa8b0ee6c9'),
      []),
  AssetType.wasmScript: Asset(
      'wasm_sequential_bd59057417018bb747b937c7a9515dc7b0af8f6807c0d6d4a2097ba0d9ec0f5d/genesis.js',
      Checksum(
          '73eef2f0d42a2bbff4bd4a865c169601e1b94a1b8057135b2aa9c2d38b3b10f9'),
      []),
};

final wasmParallel = <AssetType, Asset>{
  AssetType.wasmModule: Asset(
      'wasm_parallel_b1729fa9fb29b85a382bd00a6eaaaea8fbaebfe2f016a9d0929dd5b62e7b324e/genesis_bg.wasm',
      Checksum(
          'f3777591c522e123676db1fb353f1d57dc5ae16302fa2543c05633adae705657'),
      []),
  AssetType.wasmScript: Asset(
      'wasm_parallel_b1729fa9fb29b85a382bd00a6eaaaea8fbaebfe2f016a9d0929dd5b62e7b324e/genesis.js',
      Checksum(
          '30c20f77b0cc527346369d1327eae497441c86b9c8fb9f159eea1c7d0220b680'),
      []),
};
```

`assets_manifest.json`

```json
{
  "assets": [
    {
      "dart_enum_name": "smbertVocab",
      "url_suffix": "smbert_v0000/vocab.txt",
      "checksum": "ffb2398810fa977ee39e8a0103b07f9c2614f1f9825df53a2c99e71569e3451c",
      "fragments": []
    },
    {
      "dart_enum_name": "smbertModel",
      "url_suffix": "smbert_v0000/smbert.onnx",
      "checksum": "c07653a3cc568876071de157f0fb8ccc1a302197ddad9ead069a86140e19d913",
      "fragments": [
        {
          "url_suffix": "smbert_v0000/smbert.onnx_11MB_00",
          "checksum": "7f4d6242ef88ab6daaf1e1fa1f030345c4f4046585cd9576112aa5391f42a000"
        },
        {
          "url_suffix": "smbert_v0000/smbert.onnx_11MB_01",
          "checksum": "bd642e69234c28b3b2c25138f997505195d43a734668ce55de7adebea691ea99"
        },
        {
          "url_suffix": "smbert_v0000/smbert.onnx_11MB_02",
          "checksum": "6987dc3baddff7b742d7a481209b2a4d30097abf911fa29d50bc271dec121008"
        },
        {
          "url_suffix": "smbert_v0000/smbert.onnx_11MB_03",
          "checksum": "510ff0115aef91497fc5ef8ffcfd4c7646644d63ddc0e0635d765cf8f19b8b36"
        },
        {
          "url_suffix": "smbert_v0000/smbert.onnx_11MB_04",
          "checksum": "52e75ac8310bead0817a4948c126d2c2105a2213ec72d4f6864fad15d07d3491"
        }
      ]
    },
    {
      "dart_enum_name": "qambertVocab",
      "url_suffix": "qambert_v0001/vocab.txt",
      "checksum": "ffb2398810fa977ee39e8a0103b07f9c2614f1f9825df53a2c99e71569e3451c",
      "fragments": []
    },
    {
      "dart_enum_name": "qambertModel",
      "url_suffix": "qambert_v0001/qambert.onnx",
      "checksum": "030e7d68cc82f59640c6b989f145d6eaa7609c8615bd6ae9e7890e1a68be5b71",
      "fragments": [
        {
          "url_suffix": "qambert_v0001/qambert.onnx_11MB_00",
          "checksum": "864c75b3132c6282ff2e266ec0e51972371fae60b1678fcb90408e92eb06dfb5"
        },
        {
          "url_suffix": "qambert_v0001/qambert.onnx_11MB_01",
          "checksum": "36d3e3162e60fefd563e7a857a74d03755ed899dc79369f3083fa5ca33c9fed5"
        },
        {
          "url_suffix": "qambert_v0001/qambert.onnx_11MB_02",
          "checksum": "7db90e4fe94690593115032d2b1a30b7b76a2a6b5481cabcc7ffab6b972a76ce"
        },
        {
          "url_suffix": "qambert_v0001/qambert.onnx_11MB_03",
          "checksum": "b0c84da6e10a8ae5c0a1470e5144a4ab5dd142342c25d12968e76f72fe17c32f"
        },
        {
          "url_suffix": "qambert_v0001/qambert.onnx_11MB_04",
          "checksum": "510e9cf989dfc50fe64348d715a543006e04f9e39d75e9ba636e139137f29ac5"
        }
      ]
    },
    {
      "dart_enum_name": "ltrModel",
      "url_suffix": "ltr_v0000/ltr.binparams",
      "checksum": "f36a874ba26ea154a02070180cfcabaf191f654ecff5bb94a8472d2881d4c062",
      "fragments": []
    }
  ],
  "upload": [
    {
      "url_suffix": "smbert_v0000/vocab.txt",
      "path": "data/smbert_v0000/vocab.txt"
    },
    {
      "url_suffix": "smbert_v0000/smbert.onnx_11MB_00",
      "path": "data/chunks/smbert_v0000/smbert.onnx_11MB_00"
    },
    {
      "url_suffix": "smbert_v0000/smbert.onnx_11MB_01",
      "path": "data/chunks/smbert_v0000/smbert.onnx_11MB_01"
    },
    {
      "url_suffix": "smbert_v0000/smbert.onnx_11MB_02",
      "path": "data/chunks/smbert_v0000/smbert.onnx_11MB_02"
    },
    {
      "url_suffix": "smbert_v0000/smbert.onnx_11MB_03",
      "path": "data/chunks/smbert_v0000/smbert.onnx_11MB_03"
    },
    {
      "url_suffix": "smbert_v0000/smbert.onnx_11MB_04",
      "path": "data/chunks/smbert_v0000/smbert.onnx_11MB_04"
    },
    {
      "url_suffix": "qambert_v0001/vocab.txt",
      "path": "data/qambert_v0001/vocab.txt"
    },
    {
      "url_suffix": "qambert_v0001/qambert.onnx_11MB_00",
      "path": "data/chunks/qambert_v0001/qambert.onnx_11MB_00"
    },
    {
      "url_suffix": "qambert_v0001/qambert.onnx_11MB_01",
      "path": "data/chunks/qambert_v0001/qambert.onnx_11MB_01"
    },
    {
      "url_suffix": "qambert_v0001/qambert.onnx_11MB_02",
      "path": "data/chunks/qambert_v0001/qambert.onnx_11MB_02"
    },
    {
      "url_suffix": "qambert_v0001/qambert.onnx_11MB_03",
      "path": "data/chunks/qambert_v0001/qambert.onnx_11MB_03"
    },
    {
      "url_suffix": "qambert_v0001/qambert.onnx_11MB_04",
      "path": "data/chunks/qambert_v0001/qambert.onnx_11MB_04"
    },
    {
      "url_suffix": "ltr_v0000/ltr.binparams",
      "path": "data/ltr_v0000/ltr.binparams"
    },
    {
      "url_suffix": "wasm_bindings/genesis.js",
      "path": "bindings/dart/example/assets/wasm_bindings/genesis.js"
    },
    {
      "url_suffix": "wasm_bindings/genesis_bg.wasm",
      "path": "bindings/dart/example/assets/wasm_bindings/genesis_bg.wasm"
    },
    {
      "url_suffix": "wasm_bindings/",
      "path": "bindings/dart/example/assets/wasm_bindings/snippets"
    }
  ],
  "wasm_assets": [
    {
      "feature": "Sequential",
      "script": {
        "checksum": "fd542114545d6cd19844881b5c9329ddfbe955ce3d4bb5ac0e1e7dba9016c04b",
        "url_suffix": "wasm_bindings/genesis.js"
      },
      "module": {
        "checksum": "0d7a2b69511df6ef9cdcdf03451c8bed2d9ef2aad0bfa02fcff5ce0c5374163d",
        "url_suffix": "wasm_bindings/genesis_bg.wasm"
      }
    },
    {
      "feature": "Parallel",
      "script": {
        "url_suffix": "",
        "checksum": ""
      },
      "module": {
        "url_suffix": "",
        "checksum": ""
      }
    }
  ]
}
```
[TY-2075]: https://xainag.atlassian.net/browse/TY-2075

[TY-2077]: https://xainag.atlassian.net/browse/TY-2077
[TY-2091]: https://xainag.atlassian.net/browse/TY-2091

### CODE REVIEW

**Review by janpetschexain — COMMENTED** [review]
lgtm 👍 

> What about moving the wasmSequential and wasmParallel hash map from the common package to the web package?

good idea, they belong solely to the `web` module anyways. if it's not too cumbersome to have two `asset.dart` modules generated, then moving them is preferable for sure.

also the changelog needs to be updated due to the public api changes of `getInputData()`.

File: generate_assets_metadata.sh [hunk 1]
```diff
@@ -102,45 +104,87 @@ gen_data_assets_metadata() {
...
+    # The S3 URI must end with '/' in order to upload a directory.
+    # We don't have to add the `snippets` folder name to the snippets url suffix
+    # because s3cmd will already take the name from the snippets path.
+    local SNIPPETS_URL_SUFFIX=$WASM_VERSION/
+    local SNIPPETS_PATH=$WASM_OUT_DIR_PATH/$WASM_VERSION/snippets
+
+    add_to_upload_list "$SNIPPETS_PATH" "{\"url_suffix\": \"$SNIPPETS_URL_SUFFIX\"}"
+}
+
+# If the given asset exists (file or directory), the function will generate the
```
Comment (acrrd) [inline generate_assets_metadata.sh:170]: Why are not gibing an error instead?

**Comment by acrrd** [comment]
We should check with blue about locally disabling lints. In alternative we can have in `common` something like:

```dart
class Feature {
  final int _value;
  final String _name;
  const Feature._(this._value, this._name);
}

class MobileFeature extends Feature {
  MobileFeature._() : super._(0, 'no feature for mobile') {
    // this constructor is still usable in this file,
    // make sure no instance of this class can exists
    throw UnimplementedError('This class cannot be instanciated');
  }
}

class WebFeature extends Feature {
  const WebFeature._(int value, String name) : super._(value, name);
}

class WebFeatures {
  static const bulkMemory = WebFeature._(0, 'bulkMemory');
  static const mutableGlobals = WebFeature._(1, 'mutableGlobals');
  static const threads = WebFeature._(2, 'threads');
}
```

and have the mock `getAssets({Set<Feature> features = const {}})` using `Feature` but the mobile and web using the respective 'variant'.

File: bindings/dart/lib/package.dart [hunk 2]
```diff
@@ -6,10 +6,13 @@ export 'src/common/reranker/ai.dart'
     if (dart.library.io) 'src/mobile/reranker/ai.dart'
     if (dart.library.js) 'src/web/reranker/ai.dart' show XaynAi;
 export 'src/common/reranker/analytics.dart' show Analytics;
-export 'src/common/reranker/data_provider.dart' show Asset, AssetType, Feature;
+export 'src/common/reranker/data_provider.dart'
+    show Asset, AssetType, wasmParallel, wasmSequential;
```
[the remark this answers is not in the document]
Comment (Robert-Steiner) [reply bindings/dart/lib/package.dart]: Nice catch, I forgot to remove them 👍

File: bindings/dart/lib/src/common/reranker/data_provider.dart [hunk 3]
```diff
@@ -57,9 +57,31 @@ class Checksum {
...
+
+class WebFeature extends Feature {
+  WebFeature._() : super._();
+}
+
+// When we compare `WasmFeature`s with each other, we rely on their object identity.
+// To avoid unexpected results when comparing them, we have to make sure that a
+// `WebFeature` cannot be created by a user but that the user only uses
+// `WasmFeature`.
+class WasmFeature {
```
Comment (acrrd) [inline bindings/dart/lib/src/common/reranker/data_provider.dart:81]: I find having both `WebFeature` and `WasmFeature` a bit confusing. Moreover this are all javascript features and are not related to wasm. If the problem was the name we can move directly the static fields in `WebFeature` and have `WebFeature.bulkMemory`. wdyt?

Comment (Robert-Steiner) [reply bindings/dart/lib/src/common/reranker/data_provider.dart:81]: sounds good 👍

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,8 +8,14 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ### Changed
 
+- The `wasmParallel` `Feature` has been the replaced by the corresponding `WebFeature`s. A caller only needs to pass the set of available features to the `getAssets` function and the library automatically returns the most suitable WASM assets for the given features.
 - MAB system is not executed anymore. Only the context value is used instead to rank the documents.
 
+### Removed
+
+- Removed the optional `features` parameter in the `SetupData` constructor.
```
Comment (acrrd) [inline CHANGELOG.md:16]: This is a breaking change, correct?

**Review by acrrd — APPROVED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (rustonaut) [reply CHANGELOG.md:16]: (Yes?, Through did we ever ship a version with that parameter included?)

**Pull request merged**